src/backend/distributed/commands/foreign_constraint.c

```diff
@@ -469,23 +449,45 @@ ColumnAppearsInForeignKeyToReferenceTable(char *columnName, Oid relationId)
 
 	/* clean up scan and close system catalog */
 	systable_endscan(scanDescriptor);
-	heap_close(pgConstraint, AccessShareLock);
+	heap_close(pgConstraint, NoLock);
```

Comment:
We open the relation with `AccessShareLock`, why do we close this with `NoLock` now?

Reply:
Actually, in much contexts, it would probably cause weird edge cases to release the lock on the `pg_constraint` as we will do some further operations according to those constraints. As we are taking `AccessShareLock` and we will release it at the end of the transaction anyway, it makes sense to close the table with `NoLock`. 

```sql
/* ----------------
 *		table_close - close a table
 *
 *		If lockmode is not "NoLock", we then release the specified lock.
 *
 *		Note that it is often sensible to hold a lock beyond relation_close;
 *		in that case, the lock is released automatically at xact end.
 *		----------------
 */
void
table_close(Relation relation, LOCKMODE lockmode)
{
	relation_close(relation, lockmode);
}
```

Reply:
And if we are releasing the lock, shouldn't reads still be allowed until the end of transaction? I am not sure what problem we are fixing with this line change.

I think this is not related to this refactor and a function change. I suggest not making it here without the correctness.

Reply:
@onderkalaci  it seems that we had NoLock and AccessShareLock in previous versions as well, what do you think about this line?

Also it might be nice to write the reason we are taking a AccessShareLock and releasing with NoLock.

Reply:
> And if we are releasing the lock, shouldn't reads still be allowed until the end of transaction? I am not sure what problem we are fixing with this line change.

I missed something to wrote there, reads still would be allowed but modifications on pg_constraints won't be allowed when we do `close(NoLock)` (afaik).

As we tie those existing functions to the same low level internal function in this pr, and as some of them were doing `close(NoLock)` previously, it makes sense to me to do so in that internal function to be on the safe side.

Reply:
I think it is OK to close with `NoLock `, though could be nice to have a separate commit for that.

Reply:
I will split that change before fixing up the commits